tests/test_cases/test_cocotb/test_testfactory.py

```diff
@@ -0,0 +1,21 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+"""
+Tests of cocotb.regression.TestFactory functionality
+"""
+from cocotb.regression import TestFactory
+
+
+async def run_testfactory_test(dut, arg1, arg2, arg3):
+    assert (arg1, arg2, arg3) in [
+        ("a1v1", "a2v1", "a4v1"),
+        ("a1v2", "a2v1", "a4v1"),
+        ("a1v1", "a2v2", "a4v2"),
+        ("a1v2", "a2v2", "a4v2"),
+    ]
```

Comment:
This proves that only the expected values are generated, but doesn't prove all the expected combinations are generated.

For a test I would have this "factoried" test here save the arguments it got into a list. Then add a test definition *after* the call to `factory.generate_test()` check the list to see if all the expected values were generated.

Reply:
Okay, is there an example of something like this somewhere?

Reply:
Not that I know of. You are trailblazing :) But it's basically a module-level set that you add the arguments into and a regular cocotb test after the testfactory code that reads the set and ensures all the combos are there.

```python
args = set()

async def test_testfactory_run(dut, arg1, arg2, arg3):
    args.add((arg1, arg2, arg3))

# ... test factory stuff ...

async def test_testfactory_generates(dut):
   assert args == set(
        ("a1v1", "a2v1", "a4v1"),
        ("a1v2", "a2v1", "a4v1"),
        ("a1v1", "a2v2", "a4v2"),
        ("a1v2", "a2v2", "a4v2"),
    )

Reply:
OK, that seems to work.  I'm a bit new to writing unit tests using a testing framework in this way, so writing tests to test that the test framework is generating tests correctly is a bit of a test (yo dawg...).